File: frontend/js/views/age_based_allocation.js
// Age-based allocation — the rule-of-N equity glidepath (equity % = N − age),
// dollar split, and a 10-year-step glidepath, via /calc/age-allocation.
// Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const FIELDS = [
    ['age', 'Age', 35],
    ['rule_n', 'Rule N (100/110/120; 0 = 110)', 110],
    ['portfolio_value_usd', 'Portfolio value ($)', 250000],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const pct = (n) => Number(n).toFixed(0) + '%';

export async function renderAgeAllocation(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.aba.h1.title">// AGE-BASED ALLOCATION</span></h1>
        <p class="muted small" data-i18n="view.aba.hint.intro">
            A simple heuristic for how much to hold in stocks: equity % = N − age. The classic
            N is 100; longer lifespans pushed it to 110 or 120 for more growth. Bonds take the
            rest, and the equity share falls as you age — a built-in glidepath toward capital
            preservation. A rule of thumb, not advice. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.aba.h2.inputs">You</h2>
            <form id="aba-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.aba.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
        </div>
        <div id="aba-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#aba-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        try {
            const r = await api.calcAgeAllocation(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.aba.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#aba-result');
    const glide = r.glidepath.map((g) => `
        <tr><td>${g.age}</td><td>${pct(g.equity_pct)}</td><td>${pct(g.bond_pct)}</td></tr>
    `).join('');
    el.innerHTML = `
        <div class="chart-panel">
            <h2>${t('view.aba.h2.result', { n: r.rule_n })}</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.aba.card.equity">Equity</div>
                    <div class="value pos">${pct(r.equity_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.aba.card.bond">Bonds</div>
                    <div class="value">${pct(r.bond_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.aba.card.equity_usd">Equity $</div>
                    <div class="value">${money(r.equity_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.aba.card.bond_usd">Bonds $</div>
                    <div class="value">${money(r.bond_usd)}</div></div>
            </div>
            <table class="data-table">
                <thead><tr>
                    <th data-i18n="view.aba.col.age">Age</th>
                    <th data-i18n="view.aba.col.equity">Equity</th>
                    <th data-i18n="view.aba.col.bond">Bonds</th>
                </tr></thead>
                <tbody>${glide}</tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/age_based_allocation.rs
//! Age-based asset allocation — the "rule of N" equity glidepath.
//!
//! A simple heuristic for how much of a portfolio to hold in stocks: equity %
//! = N − age. The classic N is 100; longer lifespans pushed it to 110 or 120
//! for more growth. Bonds take the rest. As you age the equity share falls,
//! shifting toward capital preservation — a built-in glidepath.
//!
//!   * equity % = clamp(N − age, 0, 100)
//!   * bond %   = 100 − equity %
//!
//! Reports the split today, the dollar allocation, and the glidepath at
//! 10-year steps. A rule of thumb, not advice. Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct AllocationInput {
    pub age: f64,
    /// The rule constant N (100/110/120; defaults to 110 if 0).
    #[serde(default)]
    pub rule_n: f64,
    pub portfolio_value_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct GlidePoint {
    pub age: f64,
    pub equity_pct: f64,
    pub bond_pct: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct AllocationResult {
    pub rule_n: f64,
    pub equity_pct: f64,
    pub bond_pct: f64,
    pub equity_usd: f64,
    pub bond_usd: f64,
    /// Equity/bond split at the current age and 10/20/30 years out.
    pub glidepath: Vec<GlidePoint>,
}

fn equity_pct_at(rule_n: f64, age: f64) -> f64 {
    (rule_n - age).clamp(0.0, 100.0)
}

pub fn analyze(i: &AllocationInput) -> AllocationResult {
    let rule_n = if i.rule_n > 0.0 { i.rule_n } else { 110.0 };
    let portfolio = i.portfolio_value_usd.max(0.0);

    let equity_pct = equity_pct_at(rule_n, i.age);
    let bond_pct = 100.0 - equity_pct;

    let glidepath = [0.0, 10.0, 20.0, 30.0]
        .iter()
        .map(|&d| {
            let age = i.age + d;
            let eq = equity_pct_at(rule_n, age);
            GlidePoint { age, equity_pct: eq, bond_pct: 100.0 - eq }
        })
        .collect();

    AllocationResult {
        rule_n,
        equity_pct,
        bond_pct,
        equity_usd: portfolio * equity_pct / 100.0,
        bond_usd: portfolio * bond_pct / 100.0,
        glidepath,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn inp(age: f64, n: f64, pv: f64) -> AllocationInput {
        AllocationInput { age, rule_n: n, portfolio_value_usd: pv }
    }

    #[test]
    fn rule_110_age_30() {
        let r = analyze(&inp(30.0, 110.0, 100_000.0));
        assert!((r.equity_pct - 80.0).abs() < 1e-9); // 110 − 30
        assert!((r.bond_pct - 20.0).abs() < 1e-9);
    }

    #[test]
    fn rule_100_age_40() {
        let r = analyze(&inp(40.0, 100.0, 100_000.0));
        assert!((r.equity_pct - 60.0).abs() < 1e-9);
    }

    #[test]
    fn dollar_allocation() {
        let r = analyze(&inp(30.0, 110.0, 100_000.0));
        assert!((r.equity_usd - 80_000.0).abs() < 1e-6);
        assert!((r.bond_usd - 20_000.0).abs() < 1e-6);
    }

    #[test]
    fn equity_clamps_at_100_for_young() {
        // Rule 120, age 10 → 110, clamps to 100.
        let r = analyze(&inp(10.0, 120.0, 100_000.0));
        assert!((r.equity_pct - 100.0).abs() < 1e-9);
        assert!(r.bond_pct.abs() < 1e-9);
    }

    #[test]
    fn equity_clamps_at_0_for_old() {
        // Rule 100, age 110 → −10, clamps to 0.
        let r = analyze(&inp(110.0, 100.0, 100_000.0));
        assert!(r.equity_pct.abs() < 1e-9);
        assert!((r.bond_pct - 100.0).abs() < 1e-9);
    }

    #[test]
    fn glidepath_decreases_equity_with_age() {
        let r = analyze(&inp(40.0, 110.0, 100_000.0));
        assert_eq!(r.glidepath.len(), 4);
        // 70 (age40), 60 (50), 50 (60), 40 (70).
        assert!((r.glidepath[0].equity_pct - 70.0).abs() < 1e-9);
        assert!((r.glidepath[3].equity_pct - 40.0).abs() < 1e-9);
        assert!(r.glidepath[0].equity_pct > r.glidepath[3].equity_pct);
    }

    #[test]
    fn default_rule_is_110() {
        let r = analyze(&inp(50.0, 0.0, 100_000.0));
        assert!((r.rule_n - 110.0).abs() < 1e-9);
        assert!((r.equity_pct - 60.0).abs() < 1e-9); // 110 − 50
    }

    #[test]
    fn glidepath_clamps_at_zero_at_old_ages() {
        // Rule 100, age 80 → 20/10/0/0 across the decades.
        let r = analyze(&inp(80.0, 100.0, 100_000.0));
        assert!((r.glidepath[0].equity_pct - 20.0).abs() < 1e-9);
        assert!(r.glidepath[2].equity_pct.abs() < 1e-9); // age 100 → 0
        assert!(r.glidepath[3].equity_pct.abs() < 1e-9); // age 110 → clamped 0
    }
}
